File: sdk/runanywhere-swift/Sources/RunAnywhere/Core/Protocols/Component/Component.swift
import Foundation

// MARK: - Core Component Protocol

/// Base protocol that all SDK components must implement
public protocol Component: AnyObject, Sendable {
    /// Unique identifier for this component type
    static var componentType: SDKComponent { get }

    /// Current state of the component
    var state: ComponentState { get }

    /// Configuration parameters for this component
    var parameters: any ComponentInitParameters { get }

    /// Initialize the component with given parameters
    /// - Parameter parameters: Component-specific initialization parameters
    /// - Throws: If initialization fails
    func initialize(with parameters: any ComponentInitParameters) async throws

    /// Clean up and release resources
    func cleanup() async throws

    /// Check if component is ready for use
    var isReady: Bool { get }

    /// Get current health status
    func healthCheck() async -> ComponentHealth

    /// Handle state transitions
    func transitionTo(state: ComponentState) async
}

// MARK: - Component Health

/// Health status for a component
public struct ComponentHealth: Sendable {
    public let isHealthy: Bool
    public let details: String?
    public let metrics: [String: Any]?

    public init(isHealthy: Bool, details: String? = nil, metrics: [String: Any]? = nil) {
        self.isHealthy = isHealthy
        self.details = details
        self.metrics = metrics
    }
}

// MARK: - Lifecycle Management Protocol

/// Protocol for components that need lifecycle management
public protocol LifecycleManaged: Component {
    /// Called before initialization
    func willInitialize() async throws

    /// Called after successful initialization
    func didInitialize() async

    /// Called before cleanup
    func willCleanup() async

    /// Called after cleanup
    func didCleanup() async

    /// Handle memory pressure
    func handleMemoryPressure() async
}

// MARK: - Model-Based Component

/// Protocol for components that require model loading
public protocol ModelBasedComponent: Component {
    /// Model identifier
    var modelId: String? { get }

    /// Check if model is loaded
    var isModelLoaded: Bool { get }

    /// Load the model
    func loadModel(modelId: String) async throws

    /// Unload the model
    func unloadModel() async throws

    /// Get model memory usage
    func getModelMemoryUsage() async -> Int64
}

// MARK: - Service Component

/// Protocol for components that provide services
public protocol ServiceComponent: Component {
    associatedtype ServiceType

    /// Get the underlying service instance
    func getService() -> ServiceType?

    /// Create service instance
    func createService() async throws -> ServiceType
}

// MARK: - Pipeline Component

/// Protocol for components that can be part of a pipeline
public protocol PipelineComponent: Component {
    /// Process input and return output
    associatedtype Input
    associatedtype Output

    func process(_ input: Input) async throws -> Output

    /// Check if this component can connect to another
    func canConnectTo(_ component: any Component) -> Bool
}

// MARK: - Component Initialization Result

/// Result of component initialization for observability
public struct ComponentInitResult: Sendable {
    public let component: SDKComponent
    public var success: Bool
    public let duration: TimeInterval
    public let adapter: String?
    public var error: String?
    public var metadata: [String: String]

    public init(
        component: SDKComponent,
        success: Bool,
        duration: TimeInterval,
        adapter: String? = nil,
        error: String? = nil,
        metadata: [String: String] = [:]
    ) {
        self.component = component
        self.success = success
        self.duration = duration
        self.adapter = adapter
        self.error = error
        self.metadata = metadata
    }
}

// MARK: - Default Implementations

extension Component {
    /// Default health check implementation
    public func healthCheck() async -> ComponentHealth {
        ComponentHealth(isHealthy: isReady, details: "Component state: \(state.rawValue)")
    }

    /// Default isReady implementation based on state
    public var isReady: Bool {
        state == .ready
    }
}

extension LifecycleManaged {
    /// Default empty implementations for lifecycle methods
    public func willInitialize() async throws {}
    public func didInitialize() async {}
    public func willCleanup() async {}
    public func didCleanup() async {}
    public func handleMemoryPressure() async {}
}

extension ModelBasedComponent {
    /// Default model memory usage
    public func getModelMemoryUsage() async -> Int64 {
        0 // Override in implementations
    }
}

extension PipelineComponent {
    /// Default connection check
    public func canConnectTo(_ component: any Component) -> Bool {
        true // Override for specific compatibility checks
    }
}

File: sdk/runanywhere-react-native/src/Core/Protocols/Component/Component.ts
/**
 * Component.ts
 *
 * Core component protocol and related protocols
 *
 * Reference: sdk/runanywhere-swift/Sources/RunAnywhere/Core/Protocols/Component/Component.swift
 */

import type { ComponentState } from '../../Models/Common/ComponentState';
import type { SDKComponent } from '../../Models/Common/SDKComponent';
import type { ComponentInitParameters } from '../../Models/Common/ComponentInitParameters';

// ============================================================================
// Core Component Protocol
// ============================================================================

/**
 * Base protocol that all SDK components must implement
 */
export interface Component {
  /**
   * Unique identifier for this component type
   */
  readonly componentType: SDKComponent;

  /**
   * Current state of the component
   */
  readonly state: ComponentState;

  /**
   * Configuration parameters for this component
   */
  readonly parameters: ComponentInitParameters;

  /**
   * Initialize the component with given parameters
   * @param parameters - Component-specific initialization parameters
   * @throws Error if initialization fails
   */
  initialize(parameters?: ComponentInitParameters): Promise<void>;

  /**
   * Clean up and release resources
   */
  cleanup(): Promise<void>;

  /**
   * Check if component is ready for use
   */
  readonly isReady: boolean;

  /**
   * Handle state transitions
   * @param state - Target state
   */
  transitionTo(state: ComponentState): Promise<void>;
}

// ============================================================================
// Lifecycle Management Protocol
// ============================================================================

/**
 * Protocol for components that need lifecycle management
 */
export interface LifecycleManaged extends Component {
  /**
   * Called before initialization
   */
  willInitialize(): Promise<void>;

  /**
   * Called after successful initialization
   */
  didInitialize(): Promise<void>;

  /**
   * Called before cleanup
   */
  willCleanup(): Promise<void>;

  /**
   * Called after cleanup
   */
  didCleanup(): Promise<void>;

  /**
   * Handle memory pressure
   */
  handleMemoryPressure(): Promise<void>;
}

// ============================================================================
// Model-Based Component
// ============================================================================

/**
 * Protocol for components that require model loading
 */
export interface ModelBasedComponent extends Component {
  /**
   * Model identifier
   */
  readonly modelId: string | null;

  /**
   * Check if model is loaded
   */
  readonly isModelLoaded: boolean;

  /**
   * Load the model
   * @param modelId - Model identifier to load
   */
  loadModel(modelId: string): Promise<void>;

  /**
   * Unload the model
   */
  unloadModel(): Promise<void>;

  /**
   * Get model memory usage
   * @returns Memory usage in bytes
   */
  getModelMemoryUsage(): Promise<number>;
}

// ============================================================================
// Service Component
// ============================================================================

/**
 * Protocol for components that provide services
 */
export interface ServiceComponent<TService> extends Component {
  /**
   * Get the underlying service instance
   * @returns The service instance or null if not initialized
   */
  getService(): TService | null;

  /**
   * Create service instance
   * @returns The created service instance
   */
  createService(): Promise<TService>;
}

// ============================================================================
// Pipeline Component
// ============================================================================

/**
 * Protocol for components that can be part of a pipeline
 */
export interface PipelineComponent<TInput, TOutput> extends Component {
  /**
   * Process input and return output
   * @param input - Input data
   * @returns Processed output
   */
  process(input: TInput): Promise<TOutput>;

  /**
   * Check if this component can connect to another
   * @param component - Component to check compatibility with
   * @returns Whether the components can be connected
   */
  canConnectTo(component: Component): boolean;
}

// ============================================================================
// Component Initialization Result
// ============================================================================

/**
 * Result of component initialization for observability
 */
export interface ComponentInitResult {
  /**
   * Component type
   */
  readonly component: SDKComponent;

  /**
   * Whether initialization was successful
   */
  success: boolean;

  /**
   * Duration of initialization in milliseconds
   */
  readonly duration: number;

  /**
   * Adapter used (if any)
   */
  readonly adapter: string | null;

  /**
   * Error message (if failed)
   */
  error: string | null;

  /**
   * Additional metadata
   */
  metadata: Record<string, string>;
}

/**
 * Create a successful component initialization result
 */
export function createComponentInitResult(
  component: SDKComponent,
  duration: number,
  adapter?: string | null,
  metadata?: Record<string, string>
): ComponentInitResult {
  return {
    component,
    success: true,
    duration,
    adapter: adapter ?? null,
    error: null,
    metadata: metadata ?? {},
  };
}

/**
 * Create a failed component initialization result
 */
export function createFailedComponentInitResult(
  component: SDKComponent,
  duration: number,
  error: string,
  metadata?: Record<string, string>
): ComponentInitResult {
  return {
    component,
    success: false,
    duration,
    adapter: null,
    error,
    metadata: metadata ?? {},
  };
}

